File: backend/api/moldex.py
from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

from api.dependencies import get_db, get_current_user_optional
from core.models import UserORM
from db.repository import Repository
from services.avisos import textos
from utils.logger import get_logger

log = get_logger(__name__)

router = APIRouter(prefix="/moldex", tags=["Moldex"])


# ── Contrato declarado (MOLDEX-INT-013) ──────────────────────────────────────
#
# La ruta devolvía `dict[str, Any]`, así que no aparecía en el OpenAPI
# versionado y la guarda de contrato del frontend no podía cubrirla: mantenerlo
# alineado dependía de que alguien se acordara. Todos los campos científicos son
# opcionales de verdad —una corrida heredada no tiene procedencia, y RDKit puede
# no calcular un descriptor—, y ninguno se rellena con un valor por defecto:
# `null` significa «no se midió», nunca cero.


class MoldexTargetRead(BaseModel):
    pdb_id: str
    name: str | None = None
    family: str | None = None
    hotspots: list[Any] = Field(default_factory=list)
    spearman_rho: float | None = None


class MoldexMetricsRead(BaseModel):
    #: kcal/mol. Más negativo = mejor.
    affinity: float | None = None
    log_p: float | None = None
    #: Daltons.
    mw: float | None = None
    #: Å².
    tpsa: float | None = None
    #: Escala 0-100.
    score: float | None = None
    gnn_score: float | None = None
    lipinski_pass: bool | None = None
    veber_pass: bool | None = None


class MoldexProvenanceRead(BaseModel):
    """Identidad y condiciones de la corrida que produjo las métricas.

    `receptor_path` NO se publica: es una ruta local de la cuenta y EVAL-INT-008
    fijó que no se renderiza. Para decidir compatibilidad basta el hash.
    """

    task_id: str | None = None
    receptor_sha256: str | None = None
    engine_version: str | None = None
    random_seed: int | None = None
    docking_protocol: dict[str, Any] | None = None


class MoldexSealRead(BaseModel):
    certified: bool
    tx_signature: str | None = None
    certified_task_id: str | None = None
    certified_total_score: float | None = None
    #: MOLDEX-SCI-001. `None` es indeterminado, y no puede leerse como «coincide».
    matches_current_run: bool | None = None


class MoldexMoleculeRead(BaseModel):
    id: str
    name: str
    smiles: str
    smiles_hash: str
    created_at: str
    evaluated_at: str | None = None
    target: MoldexTargetRead
    metrics: MoldexMetricsRead
    provenance: MoldexProvenanceRead
    scientific_warnings: list[str] = Field(default_factory=list)
    hotspots_hit: list[str] = Field(default_factory=list)
    blockchain: MoldexSealRead


class MoldexCatalogRead(BaseModel):
    count: int
    total: int
    limit: int | None = None
    offset: int
    has_next: bool
    results: list[MoldexMoleculeRead] = Field(default_factory=list)


def _estado_del_sello(res: Any) -> dict[str, Any]:
    """Declara qué certificó el sello y si sigue describiendo la fila vigente.

    MOLDEX-SCI-001. `evaluation_results` es la proyección mutable de la última
    evaluación: al reevaluar la molécula se reescriben score, afinidad, receptor
    y semilla, pero el sello se queda. Sin esta comparación, la ficha mostraba
    la insignia CERTIFIED junto a cifras que la cadena nunca atestiguó.

    `matches_current_run` tiene tres valores y los tres significan cosas
    distintas:

    - ``True``  — el sello describe la corrida que la ficha está mostrando;
    - ``False`` — la molécula se reevaluó después de certificarla; lo sellado y
      lo mostrado son corridas distintas;
    - ``None``  — indeterminado. O no hay sello, o es anterior a v14 y no
      registra qué corrida cubrió. Suponer que coincide sería reintroducir el
      defecto, así que no se supone.
    """
    firma = res.blockchain_tx_id
    if not firma:
        return {
            "certified": False,
            "tx_signature": None,
            "certified_task_id": None,
            "certified_total_score": None,
            "matches_current_run": None,
        }

    task_sellada = res.certified_task_id
    score_sellado = res.certified_total_score

    if task_sellada is None:
        coincide = None
    else:
        # El cero medido es un valor válido, no un faltante: se compara con
        # `is not None`, nunca por veracidad (MOLDEX-SCI-002).
        coincide = task_sellada == res.task_id and score_sellado == res.total_score

    return {
        "certified": True,
        "tx_signature": firma,
        "certified_task_id": task_sellada,
        "certified_total_score": score_sellado,
        "matches_current_run": coincide,
    }

@router.get(
    "",
    response_model=MoldexCatalogRead,
    summary="Obtiene el catálogo de moléculas evaluadas (Moldex)",
)
async def get_moldex(
    target_pdb_id: str | None = Query(None, description="Filtrar por ID de PDB (ej: 7E2Y)"),
    limit: int | None = Query(
        None,
        ge=1,
        le=200,
        description="Máximo número de moléculas a devolver. Default None = todas (legacy compat).",
    ),
    offset: int = Query(0, ge=0, description="Índice a partir del cual empezar (para paginación)."),
    current_user: UserORM | None = Depends(get_current_user_optional),
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    repo = Repository(db)

    if current_user is None:
        return {
            "count": 0,
            "total": 0,
            "limit": limit,
            "offset": offset,
            "has_next": False,
            "results": [],
        }

    results, total = await repo.get_moldex_molecules(
        user_id=current_user.id,
        target_pdb_id=target_pdb_id,
        limit=limit,
        offset=offset,
    )

    # Formatear respuesta para la UI de Pokedex
    catalog = []
    for res in results:
        mol = res.molecule
        target = mol.target

        catalog.append({
            "id": str(mol.id),
            "name": mol.name or f"Ligando {mol.smiles_hash[:8]}",
            "smiles": mol.smiles,
            "smiles_hash": mol.smiles_hash,
            "created_at": mol.created_at.isoformat(),
            # Fecha de la evaluación, distinta de la creación de la molécula.
            # El cliente ya ordenaba por este campo; nunca se emitía.
            "evaluated_at": (
                res.evaluated_at.isoformat() if res.evaluated_at is not None else None
            ),
            # MOLDEX-INT-007. Sin esto una ficha no puede decir de qué corrida
            # salió su número, y ni el sello (MOLDEX-SCI-001) ni la comparación
            # (MOLDEX-SCI-004) pueden comprobar contra qué se están midiendo.
            # Todo sale de la fila que ya se cargó: ni join ni consulta extra.
            # `receptor_path` se omite a propósito — es una ruta local del
            # usuario y EVAL-INT-008 fijó que no se publica. Para decidir
            # compatibilidad basta el hash.
            "provenance": {
                "task_id": res.task_id,
                "receptor_sha256": res.receptor_sha256,
                "engine_version": res.vina_version,
                "random_seed": res.vina_random_seed,
                "docking_protocol": res.docking_protocol,
            },
            "target": {
                "pdb_id": target.pdb_id,
                "name": target.name,
                "family": target.structural_family,
                "hotspots": target.hotspots or [],
                "spearman_rho": target.spearman_rho,
            },
            "metrics": {
                "affinity": res.affinity_kcal,
                "log_p": res.log_p,
                "mw": res.molecular_weight,
                "tpsa": res.tpsa,
                "score": res.total_score,
                "gnn_score": res.gnn_score,
                "lipinski_pass": res.lipinski_pass,
                "veber_pass": res.veber_pass,
            },
            "scientific_warnings": textos(res.scientific_warnings),
            "hotspots_hit": res.hotspots_hit or [],
            "blockchain": _estado_del_sello(res),
        })

    has_next = (offset + (limit or len(catalog))) < total if limit is not None else False

    return {
        "count": len(catalog),
        "total": total,
        "limit": limit,
        "offset": offset,
        "has_next": has_next,
        "results": catalog,
    }

File: frontend/lib/moldex.ts
// =====================================================================
// MOLDEX-INT-007 — contrato tipado del catálogo Moldex
// =====================================================================
//
// `GET /moldex` devolvía `dict[str, Any]` en el backend y `Promise<any>` aquí,
// así que ninguna de las dos puntas obligaba a la otra a nada. Una ficha no
// podía decir de qué corrida salió su número, y por eso ni el sello
// (MOLDEX-SCI-001) ni la comparación (MOLDEX-SCI-004) podían comprobar contra
// qué se estaban midiendo.
//
// Nota de contrato: `/moldex` no declara `response_model` en FastAPI, así que
// no aparece en `docs/api/openapi-current.json` y la guarda automática de
// contrato (`evaluationResultContract.test.ts`) no lo cubre. Mientras siga así,
// este archivo y sus pruebas son la única definición del contrato.

/** Identidad y condiciones de la corrida que produjo las métricas de la ficha. */
export type MoldexProvenance = {
  /** Identidad durable de la corrida. `null` en evaluaciones anteriores al registro. */
  task_id: string | null;
  /** SHA-256 del receptor preparado. Decide si dos fichas son comparables. */
  receptor_sha256: string | null;
  engine_version: string | null;
  random_seed: number | null;
  docking_protocol: Record<string, unknown> | null;
};

export type MoldexTarget = {
  pdb_id: string;
  name: string | null;
  family: string | null;
  hotspots: Array<{ name?: string; [k: string]: unknown }>;
  spearman_rho: number | null;
};

/**
 * Métricas de la ficha. Todas son opcionales de verdad: RDKit puede no calcular
 * un descriptor y el pipeline puede no producir un score. Un `null` aquí
 * significa «no se midió», y no debe convertirse en 0 en ninguna vista.
 */
export type MoldexMetrics = {
  /** kcal/mol. Más negativo = mejor. */
  affinity: number | null;
  log_p: number | null;
  /** Daltons. */
  mw: number | null;
  /** Å². */
  tpsa: number | null;
  /** Escala 0-100. */
  score: number | null;
  /** RTMScore. Sólo en bases antiguas: la app de escritorio nunca lo produce. No se pinta. */
  gnn_score: number | null;
  /**
   * Salida sigmoide de la CL-GNN, en (0, 1), sin calibrar para los pesos que
   * viajan y con peso 0 en el ranking. Nulo si no corrió o falló. Opcional
   * porque un backend anterior a `bb788f5` no lo envía.
   */
  clgnn_score?: number | null;
  lipinski_pass: boolean | null;
  veber_pass: boolean | null;
};

export type MoldexMolecule = {
  id: string;
  name: string;
  smiles: string;
  smiles_hash: string;
  /** Fecha de creación de la molécula. */
  created_at: string;
  /** Fecha de la evaluación. Distinta de `created_at`. */
  evaluated_at: string | null;
  target: MoldexTarget;
  metrics: MoldexMetrics;
  provenance: MoldexProvenance;
  scientific_warnings: string[];
  hotspots_hit: string[];
  blockchain: MoldexSeal;
};

/**
 * Estado del sello de certificación.
 *
 * MOLDEX-SCI-001: el sello vive en la misma fila mutable que las métricas, así
 * que una reevaluación posterior lo deja describiendo una corrida que la ficha
 * ya no muestra.
 */
export type MoldexSeal = {
  certified: boolean;
  tx_signature: string | null;
  /** Corrida que se certificó. `null` en sellos anteriores al registro. */
  certified_task_id: string | null;
  /** Score que quedó atestiguado en la cadena. */
  certified_total_score: number | null;
  /**
   * `true` el sello describe la corrida mostrada; `false` la molécula se
   * reevaluó después de certificarla; `null` indeterminado — no hay sello, o es
   * anterior al registro y no se sabe qué cubrió. Nunca se supone que coincide.
   */
  matches_current_run: boolean | null;
};

/** Cómo debe leerse un sello en la interfaz. */
export type LecturaDelSello =
  | "sin-sello"
  | "vigente"
  | "corrida-anterior"
  | "indeterminado";

/**
 * Traduce el sello a la única afirmación que la ficha puede sostener.
 *
 * La insignia sólo puede decir «certificado» a secas cuando el sello describe la
 * corrida que se está mostrando. En los otros dos casos afirmarlo sería
 * respaldar cifras que la cadena no atestiguó.
 */
export function leerSello(sello: MoldexSeal | undefined | null): LecturaDelSello {
  if (!sello?.certified) return "sin-sello";
  if (sello.matches_current_run === true) return "vigente";
  if (sello.matches_current_run === false) return "corrida-anterior";
  return "indeterminado";
}

export type MoldexCatalog = {
  count: number;
  total: number;
  limit: number | null;
  offset: number;
  has_next: boolean;
  results: MoldexMolecule[];
};

/**
 * Marca temporal por la que se ordena «RECIENTES».
 *
 * La fecha que le importa al investigador es la de la evaluación, no la del día
 * en que dibujó la molécula. `evaluated_at` es `null` en corridas heredadas: en
 * ese caso se cae a `created_at` en vez de mandarlas al fondo.
 */
export function marcaDeOrden(molecula: Pick<MoldexMolecule, "evaluated_at" | "created_at">): number {
  const bruto = molecula.evaluated_at ?? molecula.created_at;
  if (!bruto) return 0;
  const t = new Date(bruto).getTime();
  return Number.isNaN(t) ? 0 : t;
}

/** Comparador descendente por fecha de evaluación. */
export function porEvaluacionReciente(
  a: Pick<MoldexMolecule, "evaluated_at" | "created_at">,
  b: Pick<MoldexMolecule, "evaluated_at" | "created_at">,
): number {
  return marcaDeOrden(b) - marcaDeOrden(a);
}

// =====================================================================
// MOLDEX-SCI-004 — compatibilidad antes de comparar
// =====================================================================

export type VeredictoComparacion = {
  comparable: boolean;
  /** Por qué no lo son. Vacío si lo son. Se acumulan todos, no sólo el primero. */
  motivos: string[];
};

/** Compara dos protocolos de docking por valor, campo a campo. */
function mismoProtocolo(
  a: Record<string, unknown> | null,
  b: Record<string, unknown> | null,
): boolean {
  if (a === null || b === null) return false;
  const clavesA = Object.keys(a).sort();
  const clavesB = Object.keys(b).sort();
  if (clavesA.length !== clavesB.length) return false;
  if (clavesA.some((k, i) => k !== clavesB[i])) return false;
  return clavesA.every((k) => Object.is(a[k], b[k]));
}

/**
 * Decide si dos fichas pueden restarse.
 *
 * Un score de docking sólo significa algo relativo al receptor y al protocolo
 * que lo produjeron. Comparar a través de receptores distintos —o de una
 * exhaustividad distinta, o de un docking rígido contra uno flexible— produce
 * una diferencia que parece una mejora y no lo es: la penalización torsional de
 * un ligando flexible basta para invertir el orden de dos candidatos.
 *
 * Cuando la procedencia no consta (corridas anteriores al registro) el
 * veredicto es **no comparable**. No se supone compatibilidad: no saber contra
 * qué se midió es precisamente el motivo para no restar.
 */
export function compararMoleculas(
  a: MoldexMolecule,
  b: MoldexMolecule,
): VeredictoComparacion {
  const motivos: string[] = [];

  if (a.target?.pdb_id !== b.target?.pdb_id) {
    motivos.push(
      `Son targets distintos (${a.target?.pdb_id ?? "sin target"} y ${b.target?.pdb_id ?? "sin target"}): sus afinidades no están en la misma escala.`,
    );
  }

  const hashA = a.provenance?.receptor_sha256 ?? null;
  const hashB = b.provenance?.receptor_sha256 ?? null;

  if (hashA === null || hashB === null) {
    motivos.push(
      "No consta el receptor de al menos una de las dos corridas, así que no puede comprobarse que se midieran contra la misma estructura.",
    );
  } else if (hashA !== hashB) {
    motivos.push(
      "El receptor preparado no es el mismo: cambia la caja, la protonación o la variante, y con ellas la escala del score.",
    );
  }

  const protoA = a.provenance?.docking_protocol ?? null;
  const protoB = b.provenance?.docking_protocol ?? null;

  if (protoA === null || protoB === null) {
    motivos.push(
      "No consta el protocolo de docking de al menos una de las dos corridas.",
    );
  } else if (!mismoProtocolo(protoA, protoB)) {
    motivos.push(
      "El protocolo de docking difiere (motor, exhaustividad, número de poses o conformaciones): los scores no son directamente restables.",
    );
  }

  return { comparable: motivos.length === 0, motivos };
}

// =====================================================================
// MOLDEX-SCI-003 — semántica nula en la tabla comparativa
// =====================================================================

export type FilaComparativa = {
  etiqueta: string;
  valorA: number | null;
  valorB: number | null;
  /** `null` cuando falta cualquiera de los dos lados. Nunca 0 por defecto. */
  delta: number | null;
  textoA: string;
  textoB: string;
  textoDelta: string;
};

const AUSENTE = "—";

function texto(valor: number | null): string {
  return valor === null ? AUSENTE : valor.toFixed(2);
}

/**
 * Construye una fila de la tabla comparativa conservando el faltante.
 *
 * Antes se hacía `valor ?? 0` y se restaba: una molécula sin logP calculable se
 * comparaba como si su logP fuera 0, y la diferencia contra la otra se pintaba
 * en verde o rojo. Un cero medido y un cero inventado se veían idénticos.
 */
export function filaComparativa(
  etiqueta: string,
  valorA: number | null | undefined,
  valorB: number | null | undefined,
): FilaComparativa {
  const a = valorA ?? null;
  const b = valorB ?? null;
  const delta = a === null || b === null ? null : b - a;

  return {
    etiqueta,
    valorA: a,
    valorB: b,
    delta,
    textoA: texto(a),
    textoB: texto(b),
    textoDelta:
      delta === null ? AUSENTE : `${delta > 0 ? "+" : ""}${delta.toFixed(2)}`,
  };
}

// =====================================================================
// MOLDEX-UX-008 — qué afirma la ficha, y sobre qué red
// =====================================================================

/**
 * Lo único que un sello acredita.
 *
 * Es el mismo alcance que ya declaraban `CertificationModal` y el PDF del
 * dossier. Vivía sólo ahí: la ficha, que es lo que el investigador tiene
 * delante todo el tiempo, mostraba una insignia y un encabezado «Evidencia
 * Digital» sin decir de qué es evidencia.
 */
export const ALCANCE_DEL_SELLO =
  "El registro acredita integridad y fecha de la revisión publicada. No demuestra " +
  "que la pose sea válida, ni que la conclusión científica sea correcta, ni " +
  "sustituye evidencia experimental.";

/**
 * Si la red del sello es de pruebas.
 *
 * Una red de desarrollo puede reiniciarse y su historial no es un registro
 * duradero. Ante una red desconocida se responde `true`: no saber dónde se
 * selló no puede leerse como «se selló en producción».
 */
export function esRedDePruebas(red: string | null | undefined): boolean {
  if (!red) return true;
  const normalizada = red.trim().toLowerCase();
  return normalizada !== "mainnet" && normalizada !== "mainnet-beta";
}

/** Enlace al explorador, con la red real en vez de un `cluster` fijo. */
export function urlDelExplorador(
  firma: string,
  red: string | null | undefined,
): string {
  const base = `https://explorer.solana.com/tx/${encodeURIComponent(firma)}`;
  if (!red) return base;
  const normalizada = red.trim().toLowerCase();
  if (normalizada === "mainnet" || normalizada === "mainnet-beta") return base;
  return `${base}?cluster=${encodeURIComponent(normalizada)}`;
}

/**
 * Comparador por score, con los faltantes siempre al final.
 *
 * MOLDEX-SCI-014. El orden usaba `(b.score || 0) - (a.score || 0)`, que trata
 * un score ausente como cero: en descendente lo hunde al fondo como si fuera el
 * peor, y en ascendente lo asciende a la cabeza como si fuera el mejor. No
 * tener score no es tener un score malo ni bueno — es no estar en la escala.
 *
 * El cero **medido** sí es un score válido y participa del orden con normalidad.
 */
export function porScore(
  direccion: "asc" | "desc",
): (a: Pick<MoldexMolecule, "metrics">, b: Pick<MoldexMolecule, "metrics">) => number {
  return (a, b) => {
    const sa = a.metrics?.score ?? null;
    const sb = b.metrics?.score ?? null;

    if (sa === null && sb === null) return 0;
    if (sa === null) return 1;
    if (sb === null) return -1;

    return direccion === "desc" ? sb - sa : sa - sb;
  };
}

/**
 * Cómo se enseña la CL-GNN en la ficha de Moldex.
 *
 * Es la salida sigmoide de un modelo **sin calibrar** para los pesos que viajan,
 * y pesa 0 en el ranking (canal entre sesiones, B→F-003). Por eso:
 * - dos decimales y nunca como porcentaje: un «74 %» se leería como una
 *   probabilidad calibrada, y no lo es;
 * - la condición va pegada al número («no pesa en el ranking»);
 * - un nulo es que no corrió o falló: se pinta «—» y se dice, nunca como cero.
 *
 * Devuelve la clave de la condición, no el texto: la traduce quien pinta.
 */
export function presentarClgnn(
  valor: number | null | undefined,
): { valor: string; condicion: "mx_clgnn_no_pesa" | "mx_clgnn_no_disponible" } {
  if (valor == null || !Number.isFinite(valor)) {
    return { valor: "—", condicion: "mx_clgnn_no_disponible" };
  }
  return { valor: valor.toFixed(2), condicion: "mx_clgnn_no_pesa" };
}
